### PR TITLE
[SPARK-27401][SQL] Refactoring conversion of Timestamp to/from java.sql.Timestamp

### PR DESCRIPTION
**Pull request by MaxGekk**
## What changes were proposed in this pull request?

In the PR, I propose simpler implementation of `toJavaTimestamp()`/`fromJavaTimestamp()` by reusing existing functions of `DateTimeUtils`. This will allow to:
- Simply implementation of `toJavaTimestamp()`, and handle properly negative inputs.
- Detect `Long` overflow in conversion of milliseconds (`java.sql.Timestamp`) to microseconds (Catalyst's Timestamp).

## How was this patch tested?

By existing test suites `DateTimeUtilsSuite`, `DateFunctionsSuite`, `DateExpressionsSuite` and `CastSuite`. And by new benchmark for export/import timestamps added to `DateTimeBenchmark`:

Before:
```
To/from java.sql.Timestamp:               Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
From java.sql.Timestamp                             290            335          49         17.2          58.0       1.0X
Collect longs                                      1234           1681         487          4.1         246.8       0.2X
Collect timestamps                                 1718           1755          63          2.9         343.7       0.2X
```

After:
```
To/from java.sql.Timestamp:               Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
From java.sql.Timestamp                             283            301          19         17.7          56.6       1.0X
Collect longs                                      1048           1087          36          4.8         209.6       0.3X
Collect timestamps                                 1425           1479          56          3.5         285.1       0.2X
```

### CODE REVIEW

**Comment by MaxGekk** [comment]
Failed ORC related test which generates random dates - `HiveOrcHadoopFsRelationSuite`. It seems ORC uses its own methods for converting Catalyst's days since epoch to something like `java.sql.Date` that is based on Julian calendar for old dates:
```
!== Correct Answer - 10 ==    == Spark Answer - 10 ==
 struct<index:int,col:date>   struct<index:int,col:date>
 [1,3477-08-12]               [1,3477-08-12]
 [2,7867-09-20]               [2,7867-09-20]
 [3,null]                     [3,null]
 [4,6577-03-26]               [4,6577-03-26]
 [5,8002-07-25]               [5,8002-07-25]
 [6,2154-05-10]               [6,2154-05-10]
 [7,4921-05-03]               [7,4921-05-03]
![8,0647-07-01]               [8,0647-06-28]
 [9,9422-12-25]               [9,9422-12-25]
 [10,2677-06-15]              [10,2677-06-15]
```
@dongjoon-hyun Am I right? It is strange that only this ORC test failed.

**Review by srowen — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @MaxGekk . I'll take a look today.

BTW, @srowen . Sorry, but do you mind if I ask why you approve this PR suffering from the UT failures?

**Comment by srowen** [comment]
Oh, really I mean this looks fine pending tests. If they're transient failures, fine. If they're not, obviously they need to be fixed. here I assume they aren't significant failures, just tests that need to be adjusted.

**Comment by dongjoon-hyun** [comment]
Oh, @MaxGekk . According to the last reverting commit, you are not aiming `DATE` in this PR. Are you going to file a JIRA issue for `DATE` type?

**Comment by dongjoon-hyun** [comment]
Thank you, @srowen . I see~
It seems that @MaxGekk reduced the scope of this PR first to pass the Jenkins.

**Comment by dongjoon-hyun** [comment]
Before the last reverting, I got the following result. @MaxGekk . If there was an issue, it might be an old Hive 1.2.1 issue which is fixed in Apache ORC. If needed, you can file a JIRA issue. For old Hive 1.2.1 issue, I can dig it for you but it might be difficult to fix it because it seems to be not inside both Apache Spark and Apache ORC.
- OrcHadoopFsRelationSuite passed.
- HiveOrcHadoopFsRelationSuite failed.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/LiteralGenerator.scala [hunk 1]
```diff
@@ -102,8 +102,14 @@ object LiteralGenerator {
...
-  lazy val timestampLiteralGen: Gen[Literal] =
-    for { t <- Arbitrary.arbLong.arbitrary } yield Literal.create(new Timestamp(t), TimestampType)
+  private def yearToMillis(year: Long): Long = {
+    year * 365 * 24 * 3600 * 1000L
+  }
+
+  lazy val timestampLiteralGen: Gen[Literal] = {
+    for { millis <- Gen.choose(yearToMillis(-9999), yearToMillis(10000)) }
+      yield Literal.create(new Timestamp(millis), TimestampType)
+  }
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/LiteralGenerator.scala]: @MaxGekk . In general, we don't change the test cases during refactoring in order to ensure the test coverage. I'm wondering if this is a required change due to the known limitation of the new functions? I'd make this as an another PR.

Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/LiteralGenerator.scala]: I changed the random generator for timestamps, and limit max/min values for milliseconds since epoch because the long random generator can produce milliseconds that caused Long overflow at conversion of milliseconds to microseconds. Internally as you know we store microseconds since epoch in `TimestampType`. For example, the old (current) generator can create an instance of `java.sql.Timestamp(-3948373668011580000, 570000000)`. New function `fromJavaTimestamp` calls `instantToMicros`, and inside of it we use `multiplyExact` which can detect `Long` overflow on multiplication:
```scala
def instantToMicros(instant: Instant): Long = {
  val us = Math.multiplyExact(-3948373668011580, 1000000)
    ...
}
```
Screen Shot 2019-04-07 at 08 41 30

Previous (current) implementation doesn't detect the overflow at all.

Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/LiteralGenerator.scala]: The long overflow exception caused failures in https://github.com/apache/spark/blob/1d20d13149140f53df307f47420740f45b4fa5f6/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala#L202

**Comment by MaxGekk** [comment]
> it because it seems to be not inside both Apache Spark and Apache ORC.
> - OrcHadoopFsRelationSuite passed.
> - HiveOrcHadoopFsRelationSuite failed.

Yeh, it seems there is a calendar incompatibility issue somewhere inside Hive + ORC which could cause the difference `![8,0647-07-01]               [8,0647-06-28]` for old days before 1582 (I guess). I just thought we switched to Proleptic Gregorian calendar everywhere in Spark (https://issues.apache.org/jira/browse/SPARK-26651).

In any case, I think my date (not timestamp) related changes are potentially more expensive comparing to current implementation because of conversion of `java.sql.Date` -> `java.time.LocalDate`. The conversion extracts components like `year`, `month` and etc from `java.sql.Date` that are not cheap but current implementation does time zone shifting which is not necessary too. In any case, I am going to leave Java <-> Catalyst's date conversions as is for now.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/LiteralGenerator.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/LiteralGenerator.scala]: That's the exact reason you should not put that in this PR, @MaxGekk .
Please do the improvement in another PR first. And do the refactoring PR later.

Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/LiteralGenerator.scala]: @dongjoon-hyun ok. I will open another PR. Thanks.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (pending Jenkins).
Thank you, @MaxGekk and @srowen .

File: sql/core/benchmarks/DateTimeBenchmark-results.txt [hunk 2]
```diff
@@ -414,3 +414,16 @@ to_date wholestage off                        1477 / 1479          0.7        14
...
+================================================================================================
+Conversion from/to external types
+================================================================================================
+
+Java HotSpot(TM) 64-Bit Server VM 1.8.0_202-b08 on Mac OS X 10.14.4
+Intel(R) Core(TM) i7-4850HQ CPU @ 2.30GHz
+To/from java.sql.Timestamp:               Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
+------------------------------------------------------------------------------------------------------------------------
+From java.sql.Timestamp                             309            316           8         16.2          61.9       1.0X
+Collect longs                                      1410           2747        1158          3.5         282.0       0.2X
```
Comment (dongjoon-hyun) [inline sql/core/benchmarks/DateTimeBenchmark-results.txt]: Oh, could you rerun this in the more stable environment?
`Stdev` looks too high to use as a future reference.

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by MaxGekk** [comment]
I changed only `sql/core/benchmarks/DateTimeBenchmark-results.txt` since the last success run https://github.com/apache/spark/pull/24311#issuecomment-481012862 . @dongjoon-hyun @srowen Does it make sense to re-run this again? It seems there is a permanent problem with python tests.

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by dongjoon-hyun** [comment]
Finally, it passed! Merged to master. Thank you, @MaxGekk and @srowen !
